### ISSUE TITLE
lecture: documentation

### ISSUE THREAD

**Issue by timm** [txt/se20#21]
- what : point description
- how: common use cases
- why: top-level motivation

examples good:

- Great WHY: Myers' intro to OOSD. nothing about objects for 30 pages. high level requirements. made Objects soundalike the only solution to all those issues
- how: common use cases: chapter 1 of "C". walk the. sold the language with lots of examples
- what: uml.  the great uml failure: not used

doc is an explanation. good explanations reuse patterns known to the reader. SE people have their patterns:

- koala blackboards
- lamp: (though the mean example). interesting case study, but not that since I fried it interns of LAMP, we could not journey thru decades of se developemnet
- lamp is a more general example of layers: lessons of layers (w.r.t MEAN and lamp)
- another famous layer pattern is model, view, controller (MVC)
- another famous layer patterns if data model dialog and DRY. hence the success of ruby. Ruby on Rails. its conceptual purity made it more palatable

lists of patterns:
- cognitive patterns, feature extraction (PIGE), match/select/act (first generation of AI).
- idioms . python. context manager (what is opened must be closed. html. files), iterators
- design : got
- architectural (bigger). GOV

btw,  patterns harder in OO that in functional Programmers. evidence: 23 GOF patterns, rarely changed. functional programming: 10 patterns before breakfast

great example of good docs. spatial trees
- none of the bling we mentioned before
- so much to get me going
- you may not like the following example, cause it does not reuse a pattern you've seen before. but for me, this is off the charts. so bear with me. 

finally, a note on doc and artifacts. so many artifacts. so few replicated. why? cause no on knows they exist! doc rules!

**Comment by timm** [comment]
TL;DR

Doco is effort. Hard effort

- "what" is not enough
- "how" makes people smile
- "why" makes them march
- patterns matter!

**Comment by timm** [comment]
som geratte doc genration tools

- markdown
- pandoc (that list of translators)
- mustache (more generally, its a template language). so much fun!
- Jekyll (more generally, its a site generation system)
- etc 
- etc
